### PR TITLE
MAINT: Label new PRs with relevant languages and modules

### PR DESCRIPTION
**Pull request by jjerphan**
#### Reference issue

Rework of https://github.com/scipy/scipy/pull/14677

Relates to https://github.com/scipy/scipy/pull/14709 and https://github.com/scipy/scipy/pull/14967

#### What does this implement/fix?

Maintainers and some contributors might be interested in browsing PRs based on the languages they modify or modules they modify.

The "Cython", "Fortran" and "C/C++", as well as labels for each modules exist to do so, but have to be manually added during triage work.

This commit adds a GitHub workflow to automatically label Pull Requests based on file they modify. The workflow is only run once when the PRs, allowing adapting the labels manually then if needed.

The setup is inspired from scikit-learn's, which relies on @thomasjpfan's project: [`thomasjpfan/labeler`](https://github.com/thomasjpfan/labeler).

#### Additional information

I have defined globs based on the pattern I saw in the code base.
Yet, there might be better patterns.

Also this setup can be extended to label PRs based on the sub-modules they modify.

Moreover it is possible to run a script to label existing Pull Requests.

This PR is inspired from https://github.com/scikit-learn/scikit-learn/pull/19850.

cc @tupui, @rgommers

### CODE REVIEW

**Comment by tupui** [comment]
Thanks @jjerphan. I had done that and we actually removed it after testing it out. The reason was that it was running after every commit and sometimes going agains manual updates

File: .github/workflows/labeler-module.yml
Comment (jjerphan) [inline .github/workflows/labeler-module.yml:4]: Hi @tupui,

The worflow currently runs after each PRs commit. I think It's possible to change its definition to only have it run when PRs are created, as follows:
```suggestion
name: "Pull Request Labeler"
on:
  pull_request_target:
    types: [created]
```

Comment (tupui) [reply .github/workflows/labeler-module.yml:4]: That could work then!

I suggest sending an email to the list as we had this and removed it. You could reuse the old thread we had on the topic.

As I said, me I am +1

Comment (jjerphan) [reply .github/workflows/labeler-module.yml:4]: I am searching for the thread in [SciPy-Dev mailing list archive](https://mail.python.org/archives/list/scipy-dev@python.org/) but I have not been able to identify it. Do you have a pointer?

Comment (tupui) [reply .github/workflows/labeler-module.yml:4]: Indeed me neither. It was probably just some exchanges during a meetup and on the PRs to add and remove them. Then maybe post a new message.

(We don't have a proper decision log.)

File: .github/workflows/labeler-module.yml
Comment (thomasjpfan) [inline .github/workflows/labeler-module.yml:11]: I made this fork to add a [max-labels feature](https://github.com/thomasjpfan/labeler#change-in-this-fork) which puts a upper bound on how many labels the labeler can add. If you do not need this feature, I'll go with https://github.com/actions/labeler.

Comment (tupui) [reply .github/workflows/labeler-module.yml:11]: We were using this one but had issues. See these PRs #14677 #14709 and the PR that removed it #14967

Comment (thomasjpfan) [reply .github/workflows/labeler-module.yml:11]: Ah yes, that is another feature I added to my fork: https://github.com/thomasjpfan/labeler/commit/64aeb548e93828fab1e5590c6b295e5f8ab5d67a . My fork does not label if the PR is already labeled with any of the matching labels. Matching labels are the ones configured in the yaml.

Concretely, with this PR, if someone manually labeled a PR with Cython, then my fork will do nothing. This is because `Cython` is one of the labels in `.github/labeler-file-extensions.yml`.

Comment (thomasjpfan) [reply .github/workflows/labeler-module.yml:11]: Now that I think about it, that was a time before `pull_request_target`. With `pull_request_target` set to trigger on `opened` or `created`, the label should not sync anymore and run once.

This has the downside of the labels not being in sync, but it will not override any labels. At scikit-learn, we use `opened` and that has been sufficient over the last ~2 years.

Comment (tupui) [reply .github/workflows/labeler-module.yml:11]: Ok thanks for the precisions. I believe if we have the insurance there is no sync, people would be ok to reconsider 😃 (still needs some discussion on the mailing list)

File: .github/workflows/labeler-module.yml
Comment (jjerphan) [reply .github/workflows/labeler-module.yml:4]: Is posting a message redirecting to this PR in one of the issue you mentioned in https://github.com/scipy/scipy/pull/16870#discussion_r951465958 appropriate?

Comment (tupui) [reply .github/workflows/labeler-module.yml:4]: Fine with me if you ping all people that were previously involved. We can do this here directly. I suggest that in your message you remind us all of the previous issue and how this PR solves that. Thanks for doing this!

**Comment by tupui** [comment]
We just had our community meeting and the reception was positive. Provided the action does not remove any manual update of the labels from a maintainer.

File: .github/workflows/labeler-module.yml
Comment (jjerphan) [reply .github/workflows/labeler-module.yml:4]: Committing suggestion based on the pieces of feedback given in https://github.com/scipy/scipy/pull/16870#issuecomment-1225679795.

File: .github/labeler-modules.yml
Comment (tupui) [inline .github/labeler-modules.yml:62]:
```suggestion
- scipy/weave/**/*

```

File: .github/workflows/pull-request-labeler.yml
Comment (tupui) [inline .github/workflows/pull-request-labeler.yml:26]:
```suggestion
        configuration-path: ".github/labeler-file-extensions.yml"

```

File: .github/labeler-modules.yml
Comment (tupui) [inline .github/labeler-modules.yml:1]: I would merge the two files no? Otherwise my understanding is that it would create 2 jobs, hence cost more CI which we want to limit.

**Review by tupui — COMMENTED** [review]
This looks good, thanks for the update @jjerphan. @thomasjpfan could you review this since it's your extension?

On my side I only have one comment about merging the two config files to only have a single job.

File: .github/labeler-modules.yml
Comment (thomasjpfan) [reply .github/labeler-modules.yml:1]: There should be no issue with combining them. The reason `scikit-learn` did not combine our jobs was that `max-labels` was configured differently for the modules.

XREF: https://github.com/scikit-learn/scikit-learn/blob/335e069acd7bb8baac654b69a7995a3d87275c94/.github/workflows/labeler-module.yml#L15

Comment (jjerphan) [reply .github/labeler-modules.yml]: Indeed. The files are merged by 370d13e2bf9b677b91536255dfbbb7ef4186fac5.

File: .github/workflows/labeler-module.yml
Comment (jjerphan) [reply .github/workflows/labeler-module.yml]: I just have sent an email on the SciPy-Dev mailing list as you proposed.

**Review by tupui — APPROVED** [review]
My take on this is 👍 Since I was involved in the previous work, I will let another maintainer have a look and merge if they also agree.

Again, thank you both for your help!

**Comment by jjerphan** [comment]
Following the @thomasjpfan's remark in https://github.com/scipy/scipy/pull/16870#discussion_r951461193, [`actions/labeler`](https://github.com/actions/labeler) can be used as the `max-labels` feature added in [`thomasjpfan/labeler`](https://github.com/thomasjpfan/labeler) is not used.

**Comment by tupui** [comment]
> Following the @thomasjpfan's remark in [#16870 (comment)](https://github.com/scipy/scipy/pull/16870#discussion_r951461193), [`actions/labeler`](https://github.com/actions/labeler) can be used as the `max-labels` feature added in [`thomasjpfan/labeler`](https://github.com/thomasjpfan/labeler) is not used.

Ideally yes an official action sounds nice. But as I said before, this action's maintenance is not ideal (maybe this is changing as a release was recently made). This is what we used and there was some sync problem-still not resolved, see https://github.com/actions/labeler/pull/113. So if your PR works, I prefer to go with it instead.

I would plan to merge this on Friday unless there are more comments/concerns.

**Comment by tupui** [comment]
Alright, let's try this out. Thanks everyone 🙇

**Comment by jjerphan** [comment]
Thank you!

IIRC, maintainer can run a script to update Pull Requests that were created before the merge of this PR.

@thomasjpfan: SciPy's maintainers are interested running it and if this does not necessitate too much search, do you still have the script you have used for scikit-learn to update Pull Requests?

**Comment by tupui** [comment]
Don't worry, I am not sure that would fly as we are pretty conservative and cautious.

**Comment by tupui** [comment]
@jjerphan there might be an issue with the labeller. I can see that it was never run even on PR made by maintainers where we don't need to accept workflows to run first.

**Comment by jjerphan** [comment]
Thanks for following-up.

In this case, I think it is be sensible trying to change the `pull_request_target` from `created` to something else and see if it resolves it.

It is a bit documented on [this piece of documentation about `pull_request_target`](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request_target) and on [this blog post section](https://frontside.com/blog/2020-05-26-github-actions-pull_request/#how-do-workflows-trigger-on-pull_request).

I think we can try replacing `created` with `ready_for_review`?

**Comment by tupui** [comment]
I would suggest to test this out in a personal fork.

**Pull request merged**
